src/linker/Linker.Dataflow/DynamicallyAccessedMembersTypeHierarchy.cs

```diff
@@ -47,26 +47,20 @@ public DynamicallyAccessedMembersTypeHierarchy (LinkContext context, MarkStep ma
 
 		public (DynamicallyAccessedMemberTypes annotation, bool applied) ProcessMarkedTypeForDynamicallyAccessedMembersHierarchy (TypeDefinition type)
 		{
-			// Non-interfaces must be marked already
-			Debug.Assert (type.IsInterface || _context.Annotations.IsMarked (type));
+			if (_typesInDynamicallyAccessedMembersHierarchy.TryGetValue (type, out var existingValue))
+				return existingValue;
 
 			DynamicallyAccessedMemberTypes annotation = _context.Annotations.FlowAnnotations.GetTypeAnnotation (type);
 			bool apply = false;
 
 			// We'll use the cache also as a way to detect and avoid recursion
```

Comment:
This comment should move above the very first thing in the method (which is what sort of checks for recursion). It will need to be updated as it now acts on all types, not just interfaces.